packages/validator/src/processor.ts

```diff
@@ -0,0 +1,183 @@
+import { Log, Result, Roots } from './types';
+import { execute } from './utils/action';
+import { validateParameter } from './validators/parameterValidator';
+import { validateCatch } from './validators/catchValidator';
+import { validateTemplate } from './validators/templateValidator';
+
+/**
+ * Recursion validating provided specification against template
+ * @param specs - specification that is being validated
+ * @param template - template the specification is validated against
+ * @param paramPath - string of parameters separated by ".", representing path to current specs location (empty string is root)
+ * @param nonRedundantParams - object containing all required and optional parameters that are being used
+ * @param roots - roots of specs and nonRedundantParams
+ * @param templatePath - path to current validator template file
+ * @param paramPathPrefix - in case roots are not the top layer parameters, parameter paths in messages will be prefixed with paramPathPrefix
+ * @returns errors and warnings that occurred in validation of provided specification
+ */
+export function processSpecs(
+  specs: any,
+  template: any,
+  paramPath: string[],
+  nonRedundantParams: any,
+  roots: Roots,
+  templatePath: string,
+  paramPathPrefix: string[] = []
+): Result {
+  let messages: Log[] = [];
+
+  for (const key of Object.keys(template)) {
+    if (key === '__ignore') {
+      for (const copy of Object.keys(specs)) {
+        if (!nonRedundantParams[copy]) {
+          nonRedundantParams[copy] = JSON.parse(JSON.stringify(specs[copy]));
+        }
+      }
+
+      continue;
+    }
+
+    switch (key) {
+      case '__conditions':
+        for (const condition of template[key]) {
+          messages.push(...validateCondition(specs, condition, paramPath, roots, templatePath, paramPathPrefix));
+        }
+
+        break;
+
+      case '__regexp':
+        messages.push(...validateRegexp(specs, template, [...paramPathPrefix, ...paramPath], roots.specs));
+        break;
+
+      case '__keyRegexp':
+        messages.push(...validateRegexp(specs, template, [...paramPathPrefix, ...paramPath], roots.specs, true));
+        break;
+
+      case '__maxSize':
+        if (template[key] < specs.length) {
+          messages.push(
+            logger.error(`${[...paramPathPrefix, ...paramPath].join('.')} must contain ${template[key]} or less items`)
+          );
+        }
+
+        break;
+
+      // validate array
+      case '__arrayItem':
+        // nonRedundantParams has to have the same structure as template
+        if (!nonRedundantParams) {
+          nonRedundantParams = [];
+        }
+
+        if (!paramPath.length) {
+          paramPath.push('[0]');
+        }
+
+        // validate each item in specs
+        for (let i = 0; i < specs.length; i++) {
+          nonRedundantParams.push({});
+
+          paramPath[paramPath.length - 1] = paramPath[paramPath.length - 1].replace(/\[[0-9]+\]$/, '') + `[${i}]`;
```

Comment:
I would prefer keeping regular expressions at one place (so the same regular expressions don't have to be initialized multiple times) and add a comment explaining what each regular expression does.

Reply:
I dont think thats something worth doing, as most of the regular expressions are very simple and its obvious what they are doing, there is maybe 4 of them that are a bit trickier. They also don't really repeat, so it wouldn't help with repeated initialization.

Reply:
I do agree with @amarthadan. When reading the code I have to understand what it does. If this was instead a constant imported from somewhere, than the variable name would "carry" the meaning. It will also make a refactor easier (e.g. change the format of indexing).

This is the only regex in this file, but there are many other regexes in other files.

Reply:
Added in fae0353b9739d359b219e25b9be72f72960c1dde